numpy/lib/function_base.py

```diff
@@ -370,17 +385,23 @@ def average(a, axis=None, weights=None, returned=False):
                 raise TypeError(
                     "Axis must be specified when shapes of a and weights "
                     "differ.")
-            if wgt.ndim != 1:
+            axis = _nx.normalize_axis_tuple(axis, a.ndim)
+            if len(axis) != wgt.ndim:
                 raise TypeError(
-                    "1D weights expected when shapes of a and weights differ.")
-            if wgt.shape[0] != a.shape[axis]:
-                raise ValueError(
-                    "Length of weights not compatible with specified axis.")
+                    "ndim of weights should equal len of axis.")
+            for i, ax in enumerate(axis):
+                if wgt.shape[i] != a.shape[ax]:
+                    raise ValueError(
+                        "Length of weights not compatible with specified axis.")
 
             # setup wgt to broadcast along axis
-            wgt = np.broadcast_to(wgt, (a.ndim-1)*(1,) + wgt.shape)
-            wgt = wgt.swapaxes(-1, axis)
+            exepted_shape = [1,] * a.ndim
+            for i, ax in enumerate(axis):
+                exepted_shape[ax] = wgt.shape[i]
+            wgt = wgt.reshape(exepted_shape)
 
+        if axis is not None:
+            axis = _nx.normalize_axis_tuple(axis, a.ndim)
```

Comment:
You end up calling this twice right now - perhaps just call it once earlier in the function?

Reply:
It's a little complicated here. 
1. when the shapes of `weights` and `a` differ, the `axis` must be specified.
2. when the shapes of 'weights` and  `a` equal, the `axis` can be None.
3. `normalize_axis_tuple` don't accept `None` as input.
So I can't just call it earlier, maybe I should change `if` in Line 403 to `elif`.

Reply:
Can you start with
```python
if axis is not None:
    axis = _nx.normalize_axis_tuple(axis, a.ndim)
```
and then in case 1), just add
```python
if axis is None:
    throw_the_error
```

Reply:
I have done it.